mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.td

```diff
@@ -286,10 +286,10 @@ def Polynomial_ConstantOp : Op<Polynomial_Dialect, "constant", [Pure]> {
     ```mlir
     #poly = #polynomial.int_polynomial<x**1024 - 1>
     #ring = #polynomial.ring<coefficientType=i32, coefficientModulus=65536:i32, polynomialModulus=#poly>
-    %0 = polynomial.constant #polynomial.int_polynomial<1 + x**2> : !polynomial.polynomial<#ring>
+    %0 = polynomial.constant {value=#polynomial.int_polynomial<1 + x**2>} : !polynomial.polynomial<#ring>
```

Comment:
This is a bit obnoxious: can we mimic arith.constant instead?

Also we should never let properties being printed in the attribute dictionary as a good practice, so either add prop-dict or list the attributes explicitly in the assembly format. (I should close this loop-hole in the infra, but that'll be a bad breaking change)

Reply:
I think for this I have to add a typed attribute int polynomial and float polynomial. I have this started in another PR I can port over

Reply:
Isn't the example you're updating here already typed? The value is: `#polynomial.int_polynomial<1 + x**2>`

Reply:
The SSA value out of the op is typed, but the attribute itself is not. A polynomial type requires an interpretation of that value within a particular ring. This includes the coefficient type (i32 vs i64), an optional coefficient modulus, and an optional polynomial modulus.

Because specifying the polynomial modulus on the type requires a polynomial attribute, I need to keep at least an untyped variant for that purpose. For but polynomial.constant, we can require the type on the attribute and not require the user to provide a type for the result of the op as a whole.

Reply:
Reverted the original commit and added f6276fe2d81a883676cfc24de152c0f16afdc16d which shows the typed variant. I actually like this approach much better, but couldn't figure it out how to make it work on my first attempt. Thanks for the push to get it working.

Reply:
My mistake, still some bugs to fix

Reply:
I worked on this for an hour or two and got it working, but not quite perfect.

In the last commit, the syntax is now

```mlir
%0 = polynomial.constant #polynomial.typed_int_polynomial<1 + x**2> : !polynomial.polynomial<ring=...>
```

I tried to get rid of the `#polynomial.typed_int_polynomial` to make it just

```mlir
%0 = polynomial.constant <1 + x**2> : !polynomial.polynomial<ring=...>
```

But I wasn't able to make it work. The parser infrastructure seems to make it infeasible to parse a dialect-defined attribute without the leading `#dialect_namespace.attribute` part, and when I tired to work around it in https://github.com/llvm/llvm-project/pull/92818/commits/ef17f2af7f75bd2c98c05720eb8d3f77d652ef43, I wasn't able to make it first try to parse the int polynomial, then fall back to parsing the float version on failure. It worked for the int polynomial, but failed for float_polynomial. It seemed like the parser was consuming the initial `<` token when trying to parse the int_polynomial, and then was not able to find the `<` when trying to parse the float_polynomial. I don't see a way to save and restore parser state to work around this...

If you have any alternative ideas, I'm all ears.

Reply:
If you only support int and float, you could have a short form:
```
%0 = polynomial.constant int<1 + x**2> : !polynomial.polynomial<ring=...>
```
(or any keyword instead of `int`/`float`)

Reply:
Great suggestion. That made it much easier.

Reply:
Does this LGTY?